enhancements/orange_okt_proposition.md

```diff
@@ -0,0 +1,326 @@
+
+So we built a framework (GO module) over the Operator SDK that must be updated as well each time the OperatorSDK version is upgraded (thought is is also the case of any operator based on the SDK).
+
+Our expectation now is to evaluate if this GO module presents any insightfull concepts, do we make it public and if both previous questions are positive, where is its place, out or inside the OperatorSDK ?
+
+
+### Goals
+
+The goal would be considered reached if :
+
+ - the GO operator developper is ensured to respect all golden rules by using the OperatorSDK without adding external code, and in simple manner
+ - the developper can better focus on its application logic and find with this tool many utilities commonly used in an application life cycle management (calling external API, check or control any application state, ...) 
+ - Over the common K8S resources life-cycle management, the SDK brings also a generic way to easily handle any application states (other than these states like "resources created, updated, deleted") but some points like "application started, servicing, healing, unavailable, any possible situation..."  
+
+### Non-Goals
+
+Non goals can be all that is already or should be covered by OLM.
+
+
+## Proposal
+
+### User Stories [optional]
+
+#### Story 1
+
+Here a first use-case simulation.
+
+I want to implement a Kubernetes Operator in GO and it will have to manage 3 resources (ex a Configmap, a Secret and a Deployment).
+
+I want to be aligned with best pratices, so I decide to use the OperatorSDK but I choose to use also the OKT (tell it like this) addon in order to get benefit of the enhanced Reconciliation process, a more straight forward CR status management and finalization, a result utility to trace what happens and prepare the reconciliation response (error + requeueing time).
+
+1. I create a project with the Operator SDK command as usual
+2. In my Controller, I choose an OKT Reconciler and add the Reconciliation steps function in which I can see distinctly the standard steps in which I should pass at each reconciliation event (each step below is an entry in an embracing `switch(step) { case "xxx": ... }`:
+	- **CR Checker** - once the CR is picked up by OKT and validated by webhooks (if any), I can tell here if I have something to do more on my CR
+	- **ObjectGetter** - Here I will add the code to create my 3, in memory, resources and add them to the OKT's registry of resources (still in memory). However at this stage OKT pick up the resource on the K8S Cluster so it knows if the resource in its registry has an existing peer on the Cluster (or cache).

```

Comment:
Assuming you mean that the operator registers 3 types with OKT, this sounds a lot like controller-runtime's [cache](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/cache).

Reply:
I must admit don't know lot of things about the controller-runtime's cache implementation.

At OKT level it is an array of a generic resource type, that OKT uses at the  "Mutator" and "Updater" reconciliation steps to loop on each resource and call respectively the "getHashXXX(), MutateWithInitialData(), MutateWithCR()" (for mutation) and a specific "CreateOrUpdate()" OKT method.